Lib/shutil.py

```diff
@@ -279,7 +279,7 @@ def copyfile(src, dst, *, follow_symlinks=True):
 
         # Issue 43219, raise a less confusing exception
         except IsADirectoryError as e:
-            if os.path.exists(dst):
+            if os.path.isdir(src) or os.path.isdir(dst):
```

Comment:
I'd prefer to move the open for the destination file to a separate `with` statement, and only handle the error for that case. The problem with a trailing slash is only with `dst`. It's opened in write mode, which is disallowed for a directory, so the `EISDIR` error occurs before the system even checks whether the file or directory exists. If `src` doesn't exist, then a trailing slash doesn't matter. For example:

```
>>> open('spam/', 'rb')
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
FileNotFoundError: [Errno 2] No such file or directory: 'spam/'

>>> open('spam/', 'wb')
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
IsADirectoryError: [Errno 21] Is a directory: 'spam/'
```

Reply:
I wasn't completely sure if on some OSes open() may return IsADirectory error in 'rb' mode, but I think it's not very likely and we already confirmed that's not the case for Linux, MacOS and Windows, so I agree with adding a separate with statement.

With 'isdir()' check, the logic is straightforward: if the error is 'IsADir' and the path is in fact a dir, it's ok to raise the error. But this relies on the subtle difference between open() and isdir() handling of non-existent "directory".

On the other hand it may be clearer to use exists() logic (pseudocode):

```
...
except IsADirectoryError:
  if not pathlib.exists(path):  # if it doesn't exist, it's more accurate to raise the following
     raise FileNotFoundError()
```

So I'm leaning to use the latter, does that sound good?